RecoverySpec failed

This could be related to #157 as I've seen this failure already a second time after that PR but not before.

```
[recovery of D_3] Read metadata from remote replication endpoints ...
[recovery of D_3] Read clocks from local event logs ...
[recovery of D_3] Delete invalidated snapshots at local endpoint ...
[recovery of D_3] Checking replication progress with remote endpoints ...
[ERROR] [12/03/2015 15:09:56.345] [location-akka.actor.default-dispatcher-15] [akka.tcp://location@127.0.0.1:2553/user/$b/$a] replication read failed: Ask timed out on [ActorSelection[Anchor(akka.tcp://location@127.0.0.1:2555/), Path(/user/acceptor)]] after [2000 ms]. Sender[Actor[akka://location/user/$b/$a#931430359]] sent message of type "com.rbmhtechnology.eventuate.ReplicationProtocol$ReplicationReadEnvelope".
[ERROR] [12/03/2015 15:09:56.350] [location-akka.actor.default-dispatcher-4] [akka.tcp://location@127.0.0.1:2554/user/$b/$a] replication read failed: Ask timed out on [ActorSelection[Anchor(akka.tcp://location@127.0.0.1:2555/), Path(/user/acceptor)]] after [2000 ms]. Sender[Actor[akka://location/user/$b/$a#-1416672152]] sent message of type "com.rbmhtechnology.eventuate.ReplicationProtocol$ReplicationReadEnvelope".
[ERROR] [12/03/2015 15:09:56.362] [location-akka.actor.default-dispatcher-15] [akka.tcp://location@127.0.0.1:2552/user/$b/$a] replication read failed: Ask timed out on [ActorSelection[Anchor(akka.tcp://location@127.0.0.1:2555/), Path(/user/acceptor)]] after [2000 ms]. Sender[Actor[akka://location/user/$b/$a#2140128214]] sent message of type "com.rbmhtechnology.eventuate.ReplicationProtocol$ReplicationReadEnvelope".
[recovery of D_3] Trigger update of inconsistent replication progress at B_3_L1 ...
[recovery of D_3] Trigger update of inconsistent replication progress at C_3_L1 ...
[recovery of D_3] Trigger update of inconsistent replication progress at A_3_L1 ...
[recovery of D_3] Confirm existence of consistent replication progress at B_3_L1 (1 of 3) ...
[recovery of D_3] Confirm existence of consistent replication progress at C_3_L1 (2 of 3) ...
[recovery of D_3] Confirm existence of consistent replication progress at A_3_L1 (3 of 3) ...
[info] - must repair inconsistencies of an endpoint that has lost all events *** FAILED ***
[info]   java.lang.AssertionError: assertion failed: timeout (20 seconds) during expectMsg while waiting for Set(a, b, c, d1, d)
[info]   at scala.Predef$.assert(Predef.scala:165)
[info]   at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:350)
[info]   at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:327)
[info]   at akka.testkit.TestKit.expectMsg(TestKit.scala:735)
[info]   at com.rbmhtechnology.eventuate.RecoverySpec$$anonfun$assertConvergence$1.apply(RecoverySpec.scala:94)
[info]   at com.rbmhtechnology.eventuate.RecoverySpec$$anonfun$assertConvergence$1.apply(RecoverySpec.scala:94)
[info]   at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
[info]   at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
[info]   at com.rbmhtechnology.eventuate.RecoverySpec.assertConvergence(RecoverySpec.scala:94)
[info]   at com.rbmhtechnology.eventuate.RecoverySpec$$anonfun$1$$anonfun$apply$mcV$sp$6.apply$mcV$sp(RecoverySpec.scala:159)
[info]   ...
```

--- a/src/it/scala/com/rbmhtechnology/eventuate/RecoverySpec.scala
+++ b/src/it/scala/com/rbmhtechnology/eventuate/RecoverySpec.scala
@@ -30,6 +30,7 @@ import com.rbmhtechnology.eventuate.utilities._
 import org.apache.commons.io.FileUtils
 import org.scalatest._
 
+import scala.collection.breakOut
 import scala.concurrent.duration._
 
 object RecoverySpec {
@@ -112,6 +113,26 @@ class RecoverySpec extends WordSpec with Matchers with ReplicationNodeRegistry w
 
       an [Exception] shouldBe thrownBy(endpoint.recover().await)
     }
+    "succeed normally if the endpoint was healthy" in {
+      val nodeA = node("A", Set("L1"), 2552, Set(replicationConnection(2553)))
+      val nodeB = node("B", Set("L1"), 2553, Set(replicationConnection(2552)))
+
+      nodeA.endpoint.activate()
+      nodeB.endpoint.activate()
+      val targetA = nodeA.endpoint.target("L1")
+      val targetB = nodeB.endpoint.target("L1")
+
+      write(targetA, (0 to 20).map("A" + _))
+      write(targetB, (0 to 20).map("B" + _))
+
+      Thread.sleep(50) // let some replication take place
+      nodeA.terminate().await
+      val restartedA = node("A", Set("L1"), 2552, Set(replicationConnection(2553)))
+
+      restartedA.endpoint.recover().await
+
+      assertConvergence((0 to 20).flatMap(i => Set(s"A$i", s"B$i"))(breakOut), restartedA, nodeB)
+    }
     "repair inconsistencies of an endpoint that has lost all events" in {
       val nodeA = node("A", Set("L1"), 2552, Set(replicationConnection(2555)))
       val nodeB = node("B", Set("L1"), 2553, Set(replicationConnection(2555)))
@@ -157,6 +178,33 @@ class RecoverySpec extends WordSpec with Matchers with ReplicationNodeRegistry w
       write(nodeD2.endpoint.target("L1"), List("d1"))
 
       assertConvergence(Set("a", "b", "c", "d", "d1"), nodeA, nodeB, nodeC, nodeD2)
+    }
+    "repair inconsistencies if recovery was stopped and restarted" in {
+      val nodeA = node("A", Set("L1"), 2552, Set(replicationConnection(2553)))
+      val nodeB = node("B", Set("L1"), 2553, Set(replicationConnection(2552)))
+
+      nodeA.endpoint.activate()
+      nodeB.endpoint.activate()
+      val targetA = nodeA.endpoint.target("L1")
+      val logDirA = logDirectory(targetA)
+      val targetB = nodeB.endpoint.target("L1")
+
+      write(targetA, (0 to 20).map("A" + _))
+      write(targetB, (0 to 20).map("B" + _))
+      assertConvergence((0 to 20).flatMap(i => Set(s"A$i", s"B$i"))(breakOut), nodeA, nodeB)
+
+      nodeA.terminate().await
+      FileUtils.deleteDirectory(logDirA)
+
+      val restartedA = node("A", Set("L1"), 2552, Set(replicationConnection(2553)))
+      restartedA.endpoint.recover()
+      Thread.sleep(2000) // Let some recovery take place
+      restartedA.terminate()
+
+      val restartedA2 = node("A", Set("L1"), 2552, Set(replicationConnection(2553)))
+      restartedA2.endpoint.recover().await
+
+      assertConvergence((0 to 20).flatMap(i => Set(s"A$i", s"B$i"))(breakOut), restartedA2, nodeB)
     }
     "repair inconsistencies of an endpoint that has lost all events but has been partially recovered from a storage backup" in {
       val nodeA = node("A", Set("L1"), 2552, Set(replicationConnection(2555)))

--- a/src/main/scala/com/rbmhtechnology/eventuate/Recovery.scala
+++ b/src/main/scala/com/rbmhtechnology/eventuate/Recovery.scala
@@ -24,6 +24,7 @@ import akka.util.Timeout
 
 import com.rbmhtechnology.eventuate.EventsourcingProtocol._
 import com.rbmhtechnology.eventuate.ReplicationProtocol._
+import com.rbmhtechnology.eventuate.log.EventLog
 import com.rbmhtechnology.eventuate.log.EventLogClock
 import com.typesafe.config.Config
 
@@ -50,12 +51,12 @@ private class RecoverySettings(config: Config) {
  * Represents a link between a local and remote event log that are subject to disaster recovery.
  *
  * @param logName Common name of the linked local and remote event log.
- * @param localLogId Local event log id.
  * @param localClock Local event log clock at the beginning of disaster recovery.
+ * @param localLogActor Local [[EventLog]] actor
  * @param remoteLogId Remote event log id.
- * @param remoteSequenceNr current sequence nr of the remote log
+ * @param remoteSequenceNr Current sequence nr of the remote log
  */
-private case class RecoveryLink(logName: String, localLogId: String, localClock: EventLogClock, remoteLogId: String, remoteSequenceNr: Long)
+private case class RecoveryLink(logName: String, localLogActor: ActorRef, localClock: EventLogClock, remoteLogId: String, remoteSequenceNr: Long)
 
 /**
  * Provides disaster recovery primitives.
@@ -124,7 +125,7 @@ private class Recovery(endpoint: ReplicationEndpoint) {
   def recoveryLinks(endpointInfos: Set[ReplicationEndpointInfo], clocks: Map[String, EventLogClock]) = for {
     endpointInfo <- endpointInfos
     logName <- endpoint.commonLogNames(endpointInfo)
-  } yield RecoveryLink(logName, endpoint.logId(logName), clocks(logName), endpointInfo.logId(logName), endpointInfo.logInfo(logName).get.sequenceNr)
+  } yield RecoveryLink(logName, endpoint.logs(logName), clocks(logName), endpointInfo.logId(logName), endpointInfo.logInfo(logName).get.sequenceNr)
 }
 
 /**
@@ -136,8 +137,8 @@ private class Recovery(endpoint: ReplicationEndpoint) {
  * This actor is also involved in disaster recovery and implements a state machine with the following
  * possible transitions:
  *
- *  - `initializing` -> `recoveringMetadata` -> `recoveringEvents` -> `processing` (when calling `endpoint.recover()`)
- *  - `initializing` -> `processing`                                               (when calling `endpoint.activate()`)
+ *  - `initializing` -> `recovering` -> `processing` (when calling `endpoint.recover()`)
+ *  - `initializing` -> `processing`                 (when calling `endpoint.activate()`)
  */
 private class Acceptor(endpoint: ReplicationEndpoint) extends Actor {
   import Acceptor._
@@ -150,28 +151,20 @@ private class Acceptor(endpoint: ReplicationEndpoint) extends Actor {
       context.become(processing)
     case Recover(links, promise) =>
       println(s"[recovery of ${endpoint.id}] Checking replication progress with remote endpoints ...")
-      context.become(recoveringMetadata(context.actorOf(Props(new RecoveryManager(endpoint.id, links))), promise))
+      context.become(recovering(context.actorOf(Props(new RecoveryManager(endpoint.id, links))), promise))
   }
 
-  def recoveringMetadata(recoveryManager: ActorRef, promise: Promise[Unit]): Receive = {
+  def recovering(recoveryManager: ActorRef, promise: Promise[Unit]): Receive = {
     case re: ReplicationReadEnvelope =>
       recoveryManager forward re
     case MetadataRecoveryCompleted =>
       endpoint.connectors.foreach(_.activate())
-      context.become(recoveringEvents(recoveryManager, promise) orElse processing)
-  }
-
-  def recoveringEvents(recoveryManager: ActorRef, promise: Promise[Unit]): Receive = {
-    case re: ReplicationReadResponseEnvelope =>
-      recoveryManager forward re
     case EventRecoveryCompleted =>
-      context.become(processing)
       promise.success(())
+      context.become(processing)
   }
 
   def processing: Receive = {
-    case ReplicationReadResponseEnvelope(readResult, _) =>
-      sender() forward readResult
     case ReplicationReadEnvelope(r, logName) =>
       endpoint.logs(logName) forward r
   }
@@ -204,11 +197,11 @@ private object Acceptor {
  *
  * - Metadata recovery: A [[ReplicationReadEnvelope]] from remote is responded with
  *   [[ReplicationReadSuccess]] with the current sequence nr to update remote's replication progress
- * - Event recovery: [[ReplicationReadResponseEnvelope]] responses are intercepted to detect when all
+ * - Event recovery: A [[ReplicationReadEnvelope]] from remote is intercepted to detect when all
  *   events, known to exist remotely at the beginning of recovery, are replicated and recovery is completed.
  *
  * [[RecoveryActor]]s are created to execute the disaster recovery steps for each replication link in `links`
- * and tracks the progress made by these actors. When all replication links have been processed this actor
+ * and the progress made by these actors is tracked. When all replication links have been processed this actor
  * notifies [[Acceptor]] (= parent) that recovery completed and ends itself and its childs.
  */
 private class RecoveryManager(endpointId: String, links: Set[RecoveryLink]) extends Actor {
@@ -236,8 +229,8 @@ private class RecoveryManager(endpointId: String, links: Set[RecoveryLink]) exte
   }
 
   def recoveringEvents(active: Set[RecoveryLink]): Receive = {
-    case ReplicationReadResponseEnvelope(readResult, sourceLogId) =>
-      compensators(sourceLogId) forward readResult
+    case ReplicationReadEnvelope(readRequest, _) =>
+      compensators(readRequest.targetLogId) forward readRequest
     case RecoveryStepCompleted(link) if active.contains(link) =>
       val updatedActive = active - link
       if (updatedActive.isEmpty) {
@@ -278,9 +271,9 @@ private class RecoveryActor(endpointId: String, link: RecoveryLink) extends Acto
   }
 
   def recoveringEvents: Receive = {
-    case readResult @ ReplicationReadSuccess(_, replicationProgress, _, _) =>
-      sender() forward readResult
-      if (link.remoteSequenceNr <= replicationProgress)
+    case readRequest: ReplicationRead =>
+      link.localLogActor forward readRequest
+      if (link.remoteSequenceNr < readRequest.fromSequenceNr)
         context.parent ! RecoveryStepCompleted(link)
   }
 }

--- a/src/main/scala/com/rbmhtechnology/eventuate/ReplicationEndpoint.scala
+++ b/src/main/scala/com/rbmhtechnology/eventuate/ReplicationEndpoint.scala
@@ -410,10 +410,8 @@ private class Replicator(target: ReplicationTarget, source: ReplicationSource, f
     implicit val timeout = Timeout(settings.readTimeout)
 
     (source.acceptor ? ReplicationReadEnvelope(ReplicationRead(storedReplicationProgress + 1, settings.batchSizeMax, filter, target.logId, self, currentTargetVersionVector), source.logName))
-      .mapTo[ReplicationReadResponse]
       .recover { case t => ReplicationReadFailure(t.getMessage, target.logId) }
-      .map(ReplicationReadResponseEnvelope(_, source.logId))
-      .pipeTo(target.endpoint.acceptor)
+      .pipeTo(self)
   }
 
   private def write(events: Seq[DurableEvent], replicationProgress: Long, currentSourceVersionVector: VectorTime): Unit = {

--- a/src/main/scala/com/rbmhtechnology/eventuate/ReplicationProtocol.scala
+++ b/src/main/scala/com/rbmhtechnology/eventuate/ReplicationProtocol.scala
@@ -152,17 +152,6 @@ object ReplicationProtocol {
   case class ReplicationRead(fromSequenceNr: Long, maxNumEvents: Int, filter: ReplicationFilter, targetLogId: String, replicator: ActorRef, currentTargetVersionVector: VectorTime) extends Format
 
   /**
-   * Marks all possible responses to a [[ReplicationRead]] request
-   */
-  trait ReplicationReadResponse extends Format
-
-  /**
-   * Wraps a [[ReplicationReadResponse]] and includes the id of the source log.
-   * This id is used to track replication progress per log in case of endpoint recovery.
-   */
-  case class ReplicationReadResponseEnvelope(payload: ReplicationReadResponse, sourceLogId: String)
-
-  /**
    * Success reply after a [[ReplicationRead]].
    *
    * @param events read events.
@@ -170,12 +159,12 @@ object ReplicationProtocol {
    *                            or equal to the sequence number of the last read
    *                            event (if any).
    */
-  case class ReplicationReadSuccess(events: Seq[DurableEvent], replicationProgress: Long, targetLogId: String, currentSourceVersionVector: VectorTime) extends DurableEventBatch with ReplicationReadResponse
+  case class ReplicationReadSuccess(events: Seq[DurableEvent], replicationProgress: Long, targetLogId: String, currentSourceVersionVector: VectorTime) extends DurableEventBatch with Format
 
   /**
    * Failure reply after a [[ReplicationRead]].
    */
-  case class ReplicationReadFailure(cause: String, targetLogId: String) extends ReplicationReadResponse
+  case class ReplicationReadFailure(cause: String, targetLogId: String) extends Format
 
   /**
    * Instructs an event log to batch-execute the given `writes`.

